### PR TITLE
#2829 Add robustness when target of FC/PP/Scenario reference is null

### PR DESCRIPTION
**Pull request by etraisnel2**
Fix issue where all available diagrams would be available in open diagram menu

double click now automatically opens wizard properties when target is null on FC/PP/Sc references

Fixed SWT  Leak in Capella wizard dialog

Change-Id: I7253303250f4506399932bdd903d8d95cebf0715

### CODE REVIEW

File: core/plugins/org.polarsys.capella.core.platform.sirius.ui.navigator/src/org/polarsys/capella/core/platform/sirius/ui/navigator/actions/providers/DynamicOpenRepresentationContributionItem.java [hunk 1]
```diff
@@ -146,16 +146,22 @@ private void createOpenRepresentationAction(EObject semanticElement, IMenuManage
...
-            } else if (semanticElement instanceof PhysicalPathReference) {
-                derivedElements.add(((PhysicalPathReference) semanticElement).getReferencedPhysicalPath());
-            }
-        } else if (semanticElement instanceof InteractionUse && representationDescription instanceof SequenceDiagramDescription) {
-            derivedElements.add(((InteractionUse)semanticElement).getReferencedScenario());
+        EObject targetSemanticElement = null;
+        if (representationDescription instanceof SequenceDiagramDescription) {
+          if (semanticElement instanceof AbstractCapability) {
+            derivedElements.addAll(((AbstractCapability) semanticElement).getOwnedScenarios());           
+          }else if(semanticElement instanceof InteractionUse){
```
[the remark this answers is not in the document]
Comment (etraisnel2) [reply core/plugins/org.polarsys.capella.core.platform.sirius.ui.navigator/src/org/polarsys/capella/core/platform/sirius/ui/navigator/actions/providers/DynamicOpenRepresentationContributionItem.java]: Fixed format

File: core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java [hunk 2]
```diff
@@ -1,54 +1,55 @@
...
+public class Scenario_RelatedFunctions implements IQuery {
+
+  public Scenario_RelatedFunctions() {
+    // Do nothing
+  }
+
+  /**
+   * @see org.polarsys.capella.common.helpers.query.IQuery#compute(java.lang.Object)
+   */
+  public List<Object> compute(Object object) {
```
Comment (mchoubby) [inline core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java:36]: List<AbstractFunction> would be better

File: core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java [hunk 3]
```diff
@@ -1,54 +1,55 @@
...
+    List<Object> result = new ArrayList<>();
+    if (object instanceof Scenario) {
+      Scenario af = (Scenario) object;
+      for (TimeLapse element : af.getOwnedTimeLapses().stream().filter(StateFragment.class::isInstance)
+          .collect(Collectors.toList())) {
+        StateFragment fragment = (StateFragment) element;
+        AbstractFunction absFunction = fragment.getRelatedAbstractFunction();
+        if (null != absFunction && isValidInstanceOf(absFunction)) {
```
Comment (mchoubby) [inline core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java:44]: you can remove null != absFunction

**Review by mchoubby — CHANGES_REQUESTED** [review]

File: core/plugins/org.polarsys.capella.core.preferences/src/org/polarsys/capella/core/commands/preferences/ui/sirius/DoubleClickBehaviourUtil.java [hunk 4]
```diff
@@ -39,12 +39,17 @@ private DoubleClickBehaviourUtil() {
...
+    return navigateOnDoubleClick && hasTargetSemanticElement && (source instanceof FunctionalChain
+        || source instanceof FunctionalChainReference
         || source instanceof PhysicalPath || source instanceof PhysicalPathReference || source instanceof InteractionUse
         || source instanceof Scenario);
   }
 
   public EObject getTargetSemanticElement(EObject target) {
+    if (target == null) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (etraisnel2) [reply core/plugins/org.polarsys.capella.core.preferences/src/org/polarsys/capella/core/commands/preferences/ui/sirius/DoubleClickBehaviourUtil.java]: Done

File: core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java (hunk 2 above)
Comment (etraisnel2) [reply core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java:36]: Just updated the format 
I agree with you but for homogeneity (all other queries have the same) let's keep it

File: core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java (hunk 3 above)
Comment (etraisnel2) [reply core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java]: done

File: core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java [hunk 5]
```diff
@@ -1,54 +1,55 @@
...
+
+  /**
+   * @see org.polarsys.capella.common.helpers.query.IQuery#compute(java.lang.Object)
+   */
+  public List<Object> compute(Object object) {
+    List<Object> result = new ArrayList<>();
+    if (object instanceof Scenario) {
+      Scenario af = (Scenario) object;
+      for (TimeLapse element : af.getOwnedTimeLapses().stream().filter(StateFragment.class::isInstance)
+          .collect(Collectors.toList())) {
```
[the remark this answers is not in the document]
Comment (etraisnel2) [reply core/plugins/org.polarsys.capella.core.semantic.queries/src/org/polarsys/capella/core/semantic/queries/basic/queries/Scenario_RelatedFunctions.java:41]: same, homogeneity, other queries are written this way

**Review by mchoubby — APPROVED** [review]

**Review by pdulth — APPROVED** [review]

**Pull request merged**